### PR TITLE
mysql,ddl: fix #4684, unsigned integer column length is not consistent with MySQL

### LINKED ISSUE #3725: Incorrect default integer size
```sql
CREATE TABLE t1 (c1 int unsigned , c2 char(100) not null default '');
```

MySQL:

```
mysql> show create table t1;
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                        |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `c1` int(10) unsigned DEFAULT NULL,
  `c2` char(100) NOT NULL DEFAULT ''
) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+-------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

TiDB:

```
mysql> show create table t1;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                         |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `c1` int(11) UNSIGNED DEFAULT NULL,
  `c2` char(100) NOT NULL DEFAULT ''
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #4684: default unsigned integer column length is not consistent with MySQL.
Please answer these questions before submitting your issue. Thanks!

```
mysql> create table `t1` (`c1` tinyint unsigned default null, `c2` smallint unsigned default null, `c3` mediumint unsigned default null, `c4` int unsigned default null, `c5` bigint unsigned default null);
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t1;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                           |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `c1` tinyint(3) unsigned DEFAULT NULL,
  `c2` smallint(5) unsigned DEFAULT NULL,
  `c3` mediumint(8) unsigned DEFAULT NULL,
  `c4` int(10) unsigned DEFAULT NULL,
  `c5` bigint(20) unsigned DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

TiDB:

```
mysql> create table `t1` (`c1` tinyint unsigned default null, `c2` smallint unsigned default null, `c3` mediumint unsigned default null, `c4` int unsigned default null, `c5` bigint unsigned default null);
Query OK, 0 rows affected (0.09 sec)

mysql> show create table t1;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table
                                                  |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `c1` tinyint(4) UNSIGNED DEFAULT NULL,
  `c2` smallint(6) UNSIGNED DEFAULT NULL,
  `c3` mediumint(9) UNSIGNED DEFAULT NULL,
  `c4` int(11) UNSIGNED DEFAULT NULL,
  `c5` bigint(20) UNSIGNED DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```


version:

```
mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                          |
+-----------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: 0.9.0
Git Commit Hash: e599ceec2b32b450a1e42ef49cdf5be7ea5853da
Git Branch: master
UTC Build Time: 2017-09-26 07:32:19 |
+-----------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by winkyao**
fix #4684 
fix #3725

### CODE REVIEW

**Comment by winkyao** [comment]
@zimulala @coocood @shenli PTAL

File: plan/typeinfer_test.go [hunk 1]
```diff
@@ -185,7 +185,7 @@ func (s *testPlanSuite) createTestCase4Columns() []typeInferTestCase {
 	return []typeInferTestCase{
 		{"c_bit        ", mysql.TypeBit, charset.CharsetBin, mysql.UnsignedFlag, 10, 0},
 		{"c_int_d      ", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag, 11, 0},
-		{"c_uint_d     ", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag | mysql.UnsignedFlag, 11, 0}, // TODO: Flen should be 10
+		{"c_uint_d     ", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag | mysql.UnsignedFlag, 10, 0}, // TODO: Flen should be 10
```
Comment (shenli) [inline plan/typeinfer_test.go]: Can we remove this TODO?

**Comment by shenli** [comment]
@coocood PTAL

File: plan/typeinfer_test.go (hunk 1 above)
Comment (winkyao) [reply plan/typeinfer_test.go]: The TODO comment is not referenced in this pr.

**Comment by shenli** [comment]
LGTM

File: ddl/column_test.go [hunk 2]
```diff
@@ -904,7 +904,7 @@ func (s *testColumnSuite) TestModifyColumn(c *C) {
 		err    error
 	}{
 		{"int", "bigint", nil},
-		{"int", "int unsigned", errUnsupportedModifyColumn.GenByArgs("unsigned true not match origin false")},
+		{"int", "int unsigned", errUnsupportedModifyColumn.GenByArgs("length 10 is less than origin 11")},
```
Comment (XuHuaiyu) [inline ddl/column_test.go]: this error message is consistent with MySQL?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -205,6 +205,11 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType) error {
 	defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimal(tp.Tp)
 	if tp.Flen == types.UnspecifiedLength {
 		tp.Flen = defaultFlen
+		if mysql.HasUnsignedFlag(tp.Flag) && tp.Tp != mysql.TypeLonglong && mysql.IsIntegerType(tp.Tp) {
+			// Issue #4684: unsigned integer(except bigint) is one fewer flen than signed integer
```
Comment (XuHuaiyu) [inline ddl/ddl_api.go]: why leave out `bigint` ?

Comment (XuHuaiyu) [reply ddl/ddl_api.go]: The flen of the unsigned integer(except bigint) is 1 digit shorter than signed integer because ...

**Comment by shenli** [comment]
@winkyao Please address the comment.

File: ddl/column_test.go (hunk 2 above)
Comment (winkyao) [reply ddl/column_test.go]: no, the error is only returned by TiDB because we not support this ddl request yet.

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**